src/state/localOrders.ts

```diff
@@ -0,0 +1,265 @@
+      // by local cancel tracking and notification
+      const isOrderCanceledByBackend = (orderId: string) =>
+        canceledOrderIdsInPayload.includes(orderId) &&
+        !state.localCancelOrders.some((order) => order.orderId === orderId);
+
+      const getNewCanceledOrderIds = (batch: LocalCancelAllData) => {
+        const newCanceledOrderIds = _.intersection(batch.orderIds, canceledOrderIdsInPayload);
+        return _.uniq([...(batch.canceledOrderIds ?? []), ...newCanceledOrderIds]);
+      };
+
+      return {
+        ...state,
+        localPlaceOrders: state.localPlaceOrders.map((order) =>
+          order.orderId &&
+          canceledOrderIdsInPayload.includes(order.orderId) &&
+          isOrderCanceledByBackend(order.orderId)
+            ? {
+                ...order,
+                submissionStatus: PlaceOrderStatuses.Canceled,
+              }
+            : order
+        ),
+        localCancelOrders: state.localCancelOrders.map((order) =>
+          canceledOrderIdsInPayload.includes(order.orderId)
+            ? { ...order, submissionStatus: CancelOrderStatuses.Canceled }
+            : order
+        ),
+        localCancelAlls: _.mapValues(state.localCancelAlls, (batch) => ({
+          ...batch,
+          canceledOrderIds: getNewCanceledOrderIds(batch),
+        })),
+      };
+    },
+    updateFilledOrders: (state, action: PayloadAction<SubaccountFill[]>) => {
+      const filledOrderIds = action.payload.map((fill) => fill.orderId).filter(isTruthy);
+      if (!filledOrderIds) return state;
+
+      const getFailedToCancelOrderIds = (batch: LocalCancelAllData) => {
+        const newFailedCancelOrderIds = _.intersection(batch.orderIds, filledOrderIds);
+        return _.uniq([...(batch.failedOrderIds ?? []), ...newFailedCancelOrderIds]);
+      };
+
+      return {
+        ...state,
+        localPlaceOrders: state.localPlaceOrders.map((order) =>
+          order.submissionStatus < PlaceOrderStatuses.Filled &&
+          order.orderId &&
+          filledOrderIds.includes(order.orderId)
+            ? {
+                ...order,
+                submissionStatus: PlaceOrderStatuses.Filled,
+              }
+            : order
+        ),
+        localCancelOrders: state.localCancelOrders.filter(
+          (order) => !filledOrderIds.includes(order.orderId)
+        ),
+        localCancelAlls: _.mapValues(state.localCancelAlls, (batch) => ({
+          ...batch,
+          failedOrderIds: getFailedToCancelOrderIds(batch),
+        })),
+      };
+    },
+    setLatestOrder: (state, action: PayloadAction<Nullable<SubaccountOrder>>) => {
+      const { clientId, id } = action.payload ?? {};
+      state.latestOrder = action.payload;
+
+      if (clientId) {
+        state.localPlaceOrders = state.localPlaceOrders.map((order) =>
+          order.clientId === clientId && order.submissionStatus < PlaceOrderStatuses.Placed
+            ? {
+                ...order,
+                orderId: id,
```

Comment:
hmm why do we need to replace `orderId` with `id` here? is it different from `order.orderId`?

Reply:
`localPlaceOrder` doesn't have the orderId yet! it only has the clientId from the place order payload. this is where we fill in the order id info from indexer